File: go/data_manager.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"time"
)

const DATA_DIR = "data"

type DataSubmission struct {
	StudentID      int
	SubjectName    string
	AssignmentName string
	Status         string
	Timestamp      string
}

type DataGrade struct {
	StudentID      int
	SubjectName    string
	AssignmentName string
	Score          float64
	Type           string
	Timestamp      string
}

type DataSubjectGrades struct {
	SubjectName      string
	AssignmentGrades map[int]map[string]float64
	ReportGrades     map[int]map[string]float64
}

type DataManager struct{}

func (dm *DataManager) InitDataDirectory() {
	os.MkdirAll(DATA_DIR, 0755)
}

func (dm *DataManager) GetCurrentTimestamp() string {
	return time.Now().Format("2006-01-02 15:04:05")
}

func (dm *DataManager) SaveNextUserId(nextId int) error {
	file, err := os.Create(filepath.Join(DATA_DIR, "next_id.txt"))
	if err != nil {
		return err
	}
	defer file.Close()

	file.WriteString(fmt.Sprintf("%d", nextId))
	return nil
}

func (dm *DataManager) LoadNextUserId() int {
	file, err := os.Open(filepath.Join(DATA_DIR, "next_id.txt"))
	if err != nil {
		return 1
	}
	defer file.Close()

	var nextId int
	fmt.Fscanf(file, "%d", &nextId)
	if nextId == 0 {
		return 1
	}
	return nextId
}

func (dm *DataManager) SaveUsers(users map[string]*User) error {
	file, err := os.Create(filepath.Join(DATA_DIR, "users.txt"))
	if err != nil {
		return err
	}
	defer file.Close()

	for _, user := range users {
		file.WriteString(fmt.Sprintf("%d,%s,%s,%d\n",
			user.ID, user.Name, user.PasswordHash, user.Role))
	}
	return nil
}

func (dm *DataManager) LoadUsers() map[string]*User {
	users := make(map[string]*User)

	file, err := os.Open(filepath.Join(DATA_DIR, "users.txt"))
	if err != nil {
		return users
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		parts := strings.Split(line, ",")
		if len(parts) < 4 {
			continue
		}

		id, _ := strconv.Atoi(parts[0])
		name := parts[1]
		passwordHash := parts[2]
		roleInt, _ := strconv.Atoi(parts[3])
		role := Role(roleInt)

		user := LoadUser(name, passwordHash, id, role)
		users[name] = user
	}

	return users
}

func (dm *DataManager) SaveSubjects(subjects []*Subject) error {
	file, err := os.Create(filepath.Join(DATA_DIR, "subjects.txt"))
	if err != nil {
		return err
	}
	defer file.Close()

	for _, subject := range subjects {
		file.WriteString(fmt.Sprintf("%s,%s,%d\n",
			subject.Name, subject.Code, subject.ProfessorID))
	}
	return nil
}

func (dm *DataManager) LoadSubjects() []*Subject {
	var subjects []*Subject

	file, err := os.Open(filepath.Join(DATA_DIR, "subjects.txt"))
	if err != nil {
		return subjects
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		parts := strings.Split(line, ",")
		if len(parts) < 3 {
			continue
		}

		name := parts[0]
		code := parts[1]
		profID, _ := strconv.Atoi(parts[2])

		subject := NewSubject(name, code, profID)
		subjects = append(subjects, subject)
	}

	return subjects
}

func (dm *DataManager) SaveEnrollments(enrollments map[int][]string) error {
	file, err := os.Create(filepath.Join(DATA_DIR, "enrollments.txt"))
	if err != nil {
		return err
	}
	defer file.Close()

	for studentID, subjects := range enrollments {
		line := fmt.Sprintf("%d", studentID)
		for _, subject := range subjects {
			line += "," + subject
		}
		file.WriteString(line + "\n")
	}
	return nil
}

func (dm *DataManager) LoadEnrollments() map[int][]string {
	enrollments := make(map[int][]string)

	file, err := os.Open(filepath.Join(DATA_DIR, "enrollments.txt"))
	if err != nil {
		return enrollments
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		parts := strings.Split(line, ",")
		if len(parts) < 1 {
			continue
		}

		studentID, _ := strconv.Atoi(parts[0])
		var subjects []string
		for i := 1; i < len(parts); i++ {
			subjects = append(subjects, parts[i])
		}

		enrollments[studentID] = subjects
	}

	return enrollments
}

func (dm *DataManager) SaveSubjectGrades(subjects []*Subject) error {
	file, err := os.Create(filepath.Join(DATA_DIR, "subject_grades.txt"))
	if err != nil {
		return err
	}
	defer file.Close()

	for _, subject := range subjects {
		subjectName := subject.Name
		if subjectName == "" {
			continue
		}

		// Сохраняем оценки за задания
		for studentID, assignmentGrades := range subject.AssignmentGrades {
			for assignmentName, grade := range assignmentGrades {
				file.WriteString(fmt.Sprintf("ASSIGNMENT,%s,%d,%s,%.2f\n",
					subjectName, studentID, assignmentName, grade))
			}
		}

		// Сохраняем оценки за доклады
		for studentID, reportGrades := range subject.ReportGrades {
			for reportName, grade := range reportGrades {
				file.WriteString(fmt.Sprintf("REPORT,%s,%d,%s,%.2f\n",
					subjectName, studentID, reportName, grade))
			}
		}
	}
	return nil
}

func (dm *DataManager) LoadSubjectGrades() map[string]*DataSubjectGrades {
	subjectGradesMap := make(map[string]*DataSubjectGrades)

	file, err := os.Open(filepath.Join(DATA_DIR, "subject_grades.txt"))
	if err != nil {
		return subjectGradesMap
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		parts := strings.Split(line, ",")
		if len(parts) < 5 {
			continue
		}

		typeStr := parts[0]
		subjectName := parts[1]
		studentID, _ := strconv.Atoi(parts[2])
		itemName := parts[3]
		grade, _ := strconv.ParseFloat(parts[4], 64)

		// Создаем запись для предмета, если ее нет
		if subjectGradesMap[subjectName] == nil {
			subjectGradesMap[subjectName] = &DataSubjectGrades{
				SubjectName:      subjectName,
				AssignmentGrades: make(map[int]map[string]float64),
				ReportGrades:     make(map[int]map[string]float64),
			}
		}

		if typeStr == "ASSIGNMENT" {
			if subjectGradesMap[subjectName].AssignmentGrades[studentID] == nil {
				subjectGradesMap[subjectName].AssignmentGrades[studentID] = make(map[string]float64)
			}
			subjectGradesMap[subjectName].AssignmentGrades[studentID][itemName] = grade
		} else if typeStr == "REPORT" {
			if subjectGradesMap[subjectName].ReportGrades[studentID] == nil {
				subjectGradesMap[subjectName].ReportGrades[studentID] = make(map[string]float64)
			}
			subjectGradesMap[subjectName].ReportGrades[studentID][itemName] = grade
		}
	}

	return subjectGradesMap
}

func (dm *DataManager) LoadSubmissions() []SubmissionRecord {
	var submissions []SubmissionRecord

	file, err := os.Open(filepath.Join(DATA_DIR, "submissions.txt"))
	if err != nil {
		return submissions
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		parts := strings.Split(line, ",")
		if len(parts) < 6 {
			continue
		}

		studentID, _ := strconv.Atoi(parts[0])
		submission := SubmissionRecord{
			StudentID:      studentID,
			SubjectName:    parts[1],
			AssignmentName: parts[2],
			Type:           parts[3],
			Status:         parts[4],
			Timestamp:      parts[5],
		}
		submissions = append(submissions, submission)
	}

	return submissions
}

func (dm *DataManager) LoadGrades() []GradeRecord {
	var grades []GradeRecord

	file, err := os.Open(filepath.Join(DATA_DIR, "grades.txt"))
	if err != nil {
		return grades
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		parts := strings.Split(line, ",")
		if len(parts) < 6 {
			continue
		}

		studentID, _ := strconv.Atoi(parts[0])
		score, _ := strconv.ParseFloat(parts[4], 64)

		grade := GradeRecord{
			StudentID:   studentID,
			SubjectName: parts[1],
			ItemName:    parts[2],
			Type:        parts[3],
			Score:       score,
			Timestamp:   parts[5],
		}
		grades = append(grades, grade)
	}

	return grades
}

func (dm *DataManager) SaveAssignments(assignments []*Assignment) error {
	file, err := os.Create(filepath.Join(DATA_DIR, "assignments.txt"))
	if err != nil {
		return err
	}
	defer file.Close()

	for _, assignment := range assignments {
		file.WriteString(fmt.Sprintf("%s,,%.2f,%s\n",
			assignment.Name, assignment.MaxScore, assignment.SubjectName))
	}
	return nil
}

func (dm *DataManager) LoadAssignments() []*Assignment {
	var assignments []*Assignment

	file, err := os.Open(filepath.Join(DATA_DIR, "assignments.txt"))
	if err != nil {
		return assignments
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		parts := strings.Split(line, ",")
		if len(parts) < 4 {
			continue
		}

		name := parts[0]
		maxScore, _ := strconv.ParseFloat(parts[2], 64)
		subjectName := parts[3]

		assignment := NewAssignment(name, subjectName, maxScore)
		assignments = append(assignments, assignment)
	}

	return assignments
}

func (dm *DataManager) SaveReports(reports []*Report) error {
	file, err := os.Create(filepath.Join(DATA_DIR, "reports.txt"))
	if err != nil {
		return err
	}
	defer file.Close()

	for _, report := range reports {
		completed := "0"
		if report.IsCompleted {
			completed = "1"
		}

		line := fmt.Sprintf("%s,%s,%d,%s",
			report.Topic, report.SubjectName, report.MaxParticipants, completed)

		students := report.GetSignedUpStudents()
		for _, studentID := range students {
			line += fmt.Sprintf(",%d", studentID)
		}

		file.WriteString(line + "\n")
	}
	return nil
}

func (dm *DataManager) LoadReports() []*Report {
	var reports []*Report

	file, err := os.Open(filepath.Join(DATA_DIR, "reports.txt"))
	if err != nil {
		return reports
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		parts := strings.Split(line, ",")
		if len(parts) < 4 {
			continue
		}

		topic := parts[0]
		subjectName := parts[1]
		maxParticipants, _ := strconv.Atoi(parts[2])
		completedStr := parts[3]

		report := NewReport(topic, subjectName, maxParticipants)

		if completedStr == "1" {
			report.MarkAsCompleted()
		}

		// Загружаем студентов
		for i := 4; i < len(parts); i++ {
			if parts[i] != "" {
				studentID, _ := strconv.Atoi(parts[i])
				report.AddStudent(studentID)
			}
		}

		reports = append(reports, report)
	}

	return reports
}

func (dm *DataManager) SaveSubmissions(submissions []SubmissionRecord) error {
	file, err := os.Create(filepath.Join(DATA_DIR, "submissions.txt"))
	if err != nil {
		return err
	}
	defer file.Close()

	for _, submission := range submissions {
		file.WriteString(fmt.Sprintf("%d,%s,%s,%s,%s,%s\n",
			submission.StudentID, submission.SubjectName, submission.AssignmentName,
			submission.Type, submission.Status, submission.Timestamp))
	}
	return nil
}

func (dm *DataManager) SaveGrades(grades []GradeRecord) error {
	file, err := os.Create(filepath.Join(DATA_DIR, "grades.txt"))
	if err != nil {
		return err
	}
	defer file.Close()

	for _, grade := range grades {
		file.WriteString(fmt.Sprintf("%d,%s,%s,%s,%.2f,%s\n",
			grade.StudentID, grade.SubjectName, grade.ItemName,
			grade.Type, grade.Score, grade.Timestamp))
	}
	return nil
}

func (dm *DataManager) SaveAllData(system *UniversitySystem) error {
	dm.InitDataDirectory()

	// Сохраняем пользователей
	users := make(map[string]*User)
	for name, user := range system.users {
		users[name] = user
	}
	dm.SaveUsers(users)

	// Сохраняем предметы
	dm.SaveSubjects(system.subjects)

	// Сохраняем задания
	dm.SaveAssignments(system.assignments)

	// Сохраняем доклады
	dm.SaveReports(system.reports)

	// Сохраняем записи на предметы
	dm.SaveEnrollments(system.studentEnrollments)

	// Сохраняем отправки
	dm.SaveSubmissions(system.submissions)

	// Сохраняем оценки
	dm.SaveGrades(system.grades)

	// Сохраняем оценки в предметах
	dm.SaveSubjectGrades(system.subjects)

	// Сохраняем следующий ID
	dm.SaveNextUserId(GetNextId())

	return nil
}

File: cpp/data_manager.cpp
#include "data_manager.h"
#include "user.h"
#include "student.h"
#include "professor.h"
#include "object.h"
#include <iostream>
#include <fstream>
#include <sstream>
#include <ctime>
#include <filesystem>
#include <algorithm>

using namespace std;

const string DataManager::DATA_DIR = "data";

void DataManager::initDataDirectory() {
    filesystem::create_directory(DATA_DIR);
}

void DataManager::saveNextUserId(int nextId) {
    ofstream file(DATA_DIR + "/next_id.txt");
    if (file.is_open()) {
        file << nextId;
        file.close();
    }
}

int DataManager::loadNextUserId() {
    ifstream file(DATA_DIR + "/next_id.txt");
    if (file.is_open()) {
        int nextId = 1;
        file >> nextId;
        file.close();
        return nextId;
    }
    return 1;
}

void DataManager::saveUsers(const map<string, shared_ptr<User>>& users) {
    ofstream file(DATA_DIR + "/users.txt");
    if (!file.is_open()) {
        return;
    }
    
    for (const auto& [name, user] : users) {
        file << user->getId() << "," << name << "," 
             << user->getPasswordHash() << "," 
             << static_cast<int>(user->getRole()) << "\n";
    }
    file.close();
}

void DataManager::saveSubjects(const vector<shared_ptr<Subject>>& subjects) {
    ofstream file(DATA_DIR + "/subjects.txt");
    if (!file.is_open()) {
        return;
    }
    
    for (const auto& subject : subjects) {
        file << subject->getName() << "," << subject->getCode() << "," 
             << subject->getProfessorId() << "\n";
    }
    file.close();
}

void DataManager::saveSubjectGrades(const vector<shared_ptr<Subject>>& subjects) {
    ofstream file(DATA_DIR + "/subject_grades.txt");
    if (!file.is_open()) {
        return;
    }
    
    for (const auto& subject : subjects) {
        string subjectName = subject->getName();
        if (subjectName.empty()) continue;
        
        // Сохраняем оценки за задания
        auto assignmentGrades = subject->getAllAssignmentGrades();
        for (const auto& [studentId, grades] : assignmentGrades) {
            for (const auto& [assignmentName, grade] : grades) {
                file << "ASSIGNMENT," << subjectName << "," 
                     << studentId << "," << assignmentName << "," << grade << "\n";
            }
        }
        
        // Сохраняем оценки за доклады
        auto reportGrades = subject->getAllReportGrades();
        for (const auto& [studentId, grades] : reportGrades) {
            for (const auto& [reportName, grade] : grades) {
                file << "REPORT," << subjectName << "," 
                     << studentId << "," << reportName << "," << grade << "\n";
            }
        }
    }
    file.close();
}

void DataManager::saveAssignments(const vector<shared_ptr<Assignment>>& assignments) {
    ofstream file(DATA_DIR + "/assignments.txt");
    if (!file.is_open()) {
        return;
    }
    
    for (const auto& assignment : assignments) {
        file << assignment->getName() << "," 
             << "" << "," 
             << assignment->getMaxScore() << "," << assignment->getSubjectName() << "\n";
    }
    file.close();
}

void DataManager::saveReports(const vector<shared_ptr<Report>>& reports) {
    ofstream file(DATA_DIR + "/reports.txt");
    if (!file.is_open()) {
        return;
    }
    
    for (const auto& report : reports) {
        file << report->getTopic() << "," << report->getSubjectName() << ","
             << report->getMaxParticipants() << "," 
             << (report->getIsCompleted() ? "1" : "0");
        
        auto students = report->getSignedUpStudents();
        for (int studentId : students) {
            file << "," << studentId;
        }
        file << "\n";
    }
    file.close();
}

void DataManager::saveEnrollments(const map<int, vector<string>>& studentEnrollments) {
    ofstream file(DATA_DIR + "/enrollments.txt");
    if (!file.is_open()) {
        return;
    }
    
    for (const auto& [studentId, subjects] : studentEnrollments) {
        file << studentId;
        for (const auto& subject : subjects) {
            file << "," << subject;
        }
        file << "\n";
    }
    file.close();
}

void DataManager::saveSubmissions(const vector<DataSubmission>& submissions) {
    ofstream file(DATA_DIR + "/submissions.txt");
    if (!file.is_open()) {
        return;
    }
    
    for (const auto& submission : submissions) {
        file << submission.studentId << "," << submission.subjectName << ","
             << submission.assignmentName << "," << submission.status << ","
             << submission.timestamp << "\n";
    }
    file.close();
}

void DataManager::saveGrades(const vector<DataGrade>& grades) {
    ofstream file(DATA_DIR + "/grades.txt");
    if (!file.is_open()) {
        return;
    }
    
    for (const auto& grade : grades) {
        file << grade.studentId << "," << grade.subjectName << ","
             << grade.assignmentName << "," << grade.score << ","
             << grade.type << "," << grade.timestamp << "\n";
    }
    file.close();
}

void DataManager::saveAllData(const map<string, shared_ptr<User>>& users,
                             const vector<shared_ptr<Subject>>& subjects,
                             const vector<shared_ptr<Assignment>>& assignments,
                             const vector<shared_ptr<Report>>& reports,
                             const map<int, vector<string>>& studentEnrollments,
                             const vector<DataSubmission>& submissions,
                             const vector<DataGrade>& grades) {
    initDataDirectory();
    
    saveUsers(users);
    saveSubjects(subjects);
    saveAssignments(assignments);
    saveReports(reports);
    saveEnrollments(studentEnrollments);
    saveSubmissions(submissions);
    saveGrades(grades);
    saveSubjectGrades(subjects);
    saveNextUserId(User::getNextId());
}

map<string, shared_ptr<User>> DataManager::loadUsers() {
    map<string, shared_ptr<User>> users;
    ifstream file(DATA_DIR + "/users.txt");
    if (!file.is_open()) {
        return users;
    }
    
    string line;
    while (getline(file, line)) {
        stringstream ss(line);
        string idStr, name, passwordHash, roleStr;
        
        if (getline(ss, idStr, ',') &&
            getline(ss, name, ',') &&
            getline(ss, passwordHash, ',') &&
            getline(ss, roleStr, ',')) {
            
            int id = stoi(idStr);
            User::Role role = static_cast<User::Role>(stoi(roleStr));
            
            shared_ptr<User> user;
            switch (role) {
                case User::Role::STUDENT:
                    user = make_shared<Student>(name, passwordHash, id);
                    break;
                case User::Role::PROFESSOR:
                    user = make_shared<Professor>(name, passwordHash, id);
                    break;
            }
            
            if (user) {
                users[name] = user;
            }
        }
    }
    file.close();
    return users;
}

vector<shared_ptr<Subject>> DataManager::loadSubjects() {
    vector<shared_ptr<Subject>> subjects;
    ifstream file(DATA_DIR + "/subjects.txt");
    if (!file.is_open()) {
        return subjects;
    }
    
    string line;
    while (getline(file, line)) {
        stringstream ss(line);
        string name, code, profIdStr;
        
        if (getline(ss, name, ',') &&
            getline(ss, code, ',') &&
            getline(ss, profIdStr, ',')) {
            
            int profId = stoi(profIdStr);
            subjects.push_back(make_shared<Subject>(name, code, profId));
        }
    }
    file.close();
    return subjects;
}

map<string, DataSubjectGrades> DataManager::loadSubjectGrades() {
    map<string, DataSubjectGrades> subjectGradesMap;
    ifstream file(DATA_DIR + "/subject_grades.txt");
    if (!file.is_open()) {
        return subjectGradesMap;
    }
    
    string line;
    while (getline(file, line)) {
        stringstream ss(line);
        string type, subjectName, studentIdStr, itemName, gradeStr;
        
        if (getline(ss, type, ',') &&
            getline(ss, subjectName, ',') &&
            getline(ss, studentIdStr, ',') &&
            getline(ss, itemName, ',') &&
            getline(ss, gradeStr, ',')) {
            
            int studentId = stoi(studentIdStr);
            double grade = stod(gradeStr);
            
            if (subjectGradesMap.find(subjectName) == subjectGradesMap.end()) {
                subjectGradesMap[subjectName] = DataSubjectGrades();
                subjectGradesMap[subjectName].subjectName = subjectName;
            }
            
            if (type == "ASSIGNMENT") {
                subjectGradesMap[subjectName].assignmentGrades[studentId][itemName] = grade;
            } else if (type == "REPORT") {
                subjectGradesMap[subjectName].reportGrades[studentId][itemName] = grade;
            }
        }
    }
    file.close();
    return subjectGradesMap;
}

vector<shared_ptr<Assignment>> DataManager::loadAssignments() {
    vector<shared_ptr<Assignment>> assignments;
    ifstream file(DATA_DIR + "/assignments.txt");
    if (!file.is_open()) {
        return assignments;
    }
    
    string line;
    while (getline(file, line)) {
        stringstream ss(line);
        string name, description, maxScoreStr, subjectName;
        
        if (getline(ss, name, ',') &&
            getline(ss, description, ',') &&
            getline(ss, maxScoreStr, ',') &&
            getline(ss, subjectName, ',')) {
            
            double maxScore = stod(maxScoreStr);
            assignments.push_back(make_shared<Assignment>(name, subjectName, maxScore));
        }
    }
    file.close();
    return assignments;
}

vector<shared_ptr<Report>> DataManager::loadReports() {
    vector<shared_ptr<Report>> reports;
    ifstream file(DATA_DIR + "/reports.txt");
    if (!file.is_open()) {
        return reports;
    }
    
    string line;
    while (getline(file, line)) {
        stringstream ss(line);
        string topic, subjectName, maxPartStr, completedStr;
        
        if (getline(ss, topic, ',') &&
            getline(ss, subjectName, ',') &&
            getline(ss, maxPartStr, ',') &&
            getline(ss, completedStr, ',')) {
            
            int maxParticipants = stoi(maxPartStr);
            auto report = make_shared<Report>(topic, subjectName, maxParticipants);
            
            string studentIdStr;
            while (getline(ss, studentIdStr, ',')) {
                if (!studentIdStr.empty()) {
                    int studentId = stoi(studentIdStr);
                    report->addStudent(studentId);
                }
            }
            
            if (completedStr == "1") {
                report->markAsCompleted();
            }
            
            reports.push_back(report);
        }
    }
    file.close();
    return reports;
}

map<int, vector<string>> DataManager::loadEnrollments() {
    map<int, vector<string>> enrollments;
    ifstream file(DATA_DIR + "/enrollments.txt");
    if (!file.is_open()) {
        return enrollments;
    }
    
    string line;
    while (getline(file, line)) {
        stringstream ss(line);
        string studentIdStr;
        
        if (getline(ss, studentIdStr, ',')) {
            int studentId = stoi(studentIdStr);
            vector<string> subjects;
            string subjectName;
            
            while (getline(ss, subjectName, ',')) {
                subjects.push_back(subjectName);
            }
            
            enrollments[studentId] = subjects;
        }
    }
    file.close();
    return enrollments;
}

vector<DataSubmission> DataManager::loadSubmissions() {
    vector<DataSubmission> submissions;
    ifstream file(DATA_DIR + "/submissions.txt");
    if (!file.is_open()) {
        return submissions;
    }
    
    string line;
    while (getline(file, line)) {
        stringstream ss(line);
        DataSubmission submission;
        string studentIdStr;
        
        if (getline(ss, studentIdStr, ',') &&
            getline(ss, submission.subjectName, ',') &&
            getline(ss, submission.assignmentName, ',') &&
            getline(ss, submission.status, ',') &&
            getline(ss, submission.timestamp, ',')) {
            
            submission.studentId = stoi(studentIdStr);
            submissions.push_back(submission);
        }
    }
    file.close();
    return submissions;
}

vector<DataGrade> DataManager::loadGrades() {
    vector<DataGrade> grades;
    ifstream file(DATA_DIR + "/grades.txt");
    if (!file.is_open()) {
        return grades;
    }
    
    string line;
    while (getline(file, line)) {
        stringstream ss(line);
        DataGrade grade;
        string studentIdStr, scoreStr;
        
        if (getline(ss, studentIdStr, ',') &&
            getline(ss, grade.subjectName, ',') &&
            getline(ss, grade.assignmentName, ',') &&
            getline(ss, scoreStr, ',') &&
            getline(ss, grade.type, ',') &&
            getline(ss, grade.timestamp, ',')) {
            
            grade.studentId = stoi(studentIdStr);
            grade.score = stod(scoreStr);
            grades.push_back(grade);
        }
    }
    file.close();
    return grades;
}

string DataManager::getCurrentTimestamp() {
    time_t now = time(nullptr);
    tm* tm = localtime(&now);
    char buffer[20];
    strftime(buffer, sizeof(buffer), "%Y-%m-%d %H:%M:%S", tm);
    return string(buffer);
}